### PR TITLE
Add duotone to image block

### PR DESCRIPTION
**Pull request by ajlende**
## Description

This is a subset of #26361 focusing just on the image block since it doesn't have any additional dependencies.

See #26752 for adding duotone as a "block supports" feature.

- Add duotone to theme.json
- Add duotone filter component
- Add `disableInserter` and `disableAlpha` props to CustomGradientBar
- Add duotone toolbar
- Add duotone to image block 

The effect is applied as an SVG filter which is [supported all the way back to IE 10](https://www.caniuse.com/svg-filters). Since it's working with SVG, a server-rendered component was required. The SVG is hidden with inline styles, and a stylesheet is written to apply the filter to a specific component using a unique (to the filter) classname.

## How has this been tested?

1. Create image block
2. Add custom duotone filter
3. Add preset duotone filter
4. Test keyboard accessibility

## Screenshots 

https://user-images.githubusercontent.com/5129775/107701409-f6e62100-6c5c-11eb-824b-6226ae499f5b.mp4

<details>
<summary>Old Screenshots</summary>
<p>

image-duotone-2020-11-05

</p>
</details>

## Types of changes

New feature

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

File: lib/duotone-filter.php [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+ * @return string           Duotone stylesheet and SVG.
+ */
+function gutenberg_render_duotone_filter( $selector, $id, $values ) {
+	ob_start();
+
+	?>
+
+	<style>
+		<?php echo $selector; ?> {
+			filter: url( <?php echo '#' . $id; ?> );
```
Comment (mkaz) [inline lib/duotone-filter.php:23]: I think you will want to use `esc_url` like so

```suggestion
			filter: url( <?php echo esc_url('#' . $id); ?> );
```

File: lib/duotone-filter.php [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+		xmlns:xlink="http://www.w3.org/1999/xlink"
+		viewBox="0 0 0 0"
+		width="0"
+		height="0"
+		focusable="false"
+		role="none"
+		style="visibility: hidden; position: absolute; left: -9999px; overflow: hidden;"
+	>
+		<defs>
+			<filter id="<?php echo $id; ?>">
```
Comment (mkaz) [inline lib/duotone-filter.php:37]: Here you'll want `esc_attr`

```suggestion
			<filter id="<?php echo esc_attr( $id ); ?>">
```

File: lib/duotone-filter.php [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+					values=".299 .587 .114 0 0
+							.299 .587 .114 0 0
+							.299 .587 .114 0 0
+							0 0 0 1 0"
+					<?php // phpcs:enable Generic.WhiteSpace.DisallowSpaceIndent ?>
+				/>
+				<feComponentTransfer color-interpolation-filters="sRGB" >
+					<feFuncR type="table" tableValues="<?php echo implode( ' ', $values['r'] ); ?>" />
+					<feFuncG type="table" tableValues="<?php echo implode( ' ', $values['g'] ); ?>" />
+					<feFuncB type="table" tableValues="<?php echo implode( ' ', $values['b'] ); ?>" />
```
Comment (mkaz) [inline lib/duotone-filter.php:50]: You'll want `esc_attr()` here:

```suggestion
					<feFuncR type="table" tableValues="<?php echo esc_attr( implode( ' ', $values['r'] ) ); ?>" />
					<feFuncG type="table" tableValues="<?php echo esc_attr( implode( ' ', $values['g'] ) ); ?>" />
					<feFuncB type="table" tableValues="<?php echo esc_attr( implode( ' ', $values['b'] ) ); ?>" />
```

**Review by mkaz — CHANGES_REQUESTED** [review]
This tested well and I didn't come across any issues during testing, it all seems to make sense and no issues with any of the additional components in various testing scenarios.

I made a few suggestions for the PHP written to screen needs some extra escaping, see [Data Sanitization documentation](https://developer.wordpress.org/themes/theme-security/data-sanitization-escaping/#escaping-securing-output) for reference.

**Review by mkaz — APPROVED** [review]
Updates for escaping attributes looks good. :+1:

I did further testing and didn't turn anything up.  It is looking good to me.

@youknowriad do you want to give it a once over?

File: lib/duotone-filter.php [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+/**
+ * SVG and stylesheet needed for rendering the duotone filter.
+ *
+ * @param  string $selector CSS selectors to apply the duotone to.
+ * @param  string $id       Unique slug for this duotone filter.
+ * @param  array  $values   R, G, and B values to filter with.
+ * @return string           Duotone stylesheet and SVG.
+ */
+function gutenberg_render_duotone_filter( $selector, $id, $values ) {
```
Comment (youknowriad) [inline lib/duotone-filter.php:16]: Why is this thing not absorbed in the block support?

Comment (ajlende) [reply lib/duotone-filter.php:16]: It's needed later in the [image block render](https://github.com/WordPress/gutenberg/pull/26751/files/912b7d03348a0f2d747102177492be5c74eca72a#diff-1e4aa29ea183efda9c795271dc1acf6e8049355db104d0dca5b0ea5caf7f619cR29) and in any block that wants to implement duotone without block supports

File: packages/block-library/src/image/edit.js [hunk 5]
```diff
@@ -275,6 +276,7 @@ export function ImageEdit( {
 	);
 
 	const classes = classnames( className, {
+		[ duotone?.id ]: duotone,
```
Comment (youknowriad) [inline packages/block-library/src/image/edit.js:279]: In #29335 I need a similar unique id to apply inline style for the current block, Do you think we can build a generic thing for that and maybe extract it to its own PR first. It could be a support flag on its own or something.

File: packages/block-library/src/image/image.js [hunk 6]
```diff
@@ -552,6 +572,13 @@ export default function Image( {
...
+			{ duotone && (
+				<DuotoneFilter
+					selector={ `.wp-block-image.${ duotone.id } img` }
+					id={ duotone.id }
+					values={ duotone.values }
+				/>
+			) }
 			{ ( ! RichText.isEmpty( caption ) || isSelected ) && (
 				<RichText
 					ref={ captionRef }
```
Comment (youknowriad) [inline packages/block-library/src/image/image.js:584]: Is there a way to absorb the changes made  here in the support flag, meaning if we want to add support to other blocks, there's no changes that should be made in the block itself?

File: packages/components/src/index.js [hunk 7]
```diff
@@ -29,6 +29,7 @@ export { default as CardFooter } from './card/footer';
 export { default as CardHeader } from './card/header';
 export { default as CardMedia } from './card/media';
 export { default as CheckboxControl } from './checkbox-control';
+export { default as CircularOptionPicker } from './circular-option-picker';
```
Comment (youknowriad) [inline packages/components/src/index.js:32]: Why is this (and the bar) exported here? Maybe instead oof exporting this low level component, we could instead add a generic duotone picker UI component here?

File: lib/duotone-filter.php (hunk 4 above)
Comment (youknowriad) [reply lib/duotone-filter.php:16]: I was thinking that ideally, all blocks should rely on the block support and don't implement this adhoc. Is this possible?

File: packages/block-library/src/image/image.js (hunk 6 above)
Comment (ajlende) [reply packages/block-library/src/image/image.js:584]: The image block was added separately so the positioning of the toolbar could be controlled. @jasmussen and @mtias said they wanted the duotone toolbar positioned alongside the crop controls or as part of the crop controls which couldn't be done with block supports (See https://github.com/WordPress/gutenberg/pull/26361#issuecomment-714415453 and the following few comments)

Comment (youknowriad) [reply packages/block-library/src/image/image.js:584]: >  as part of the crop controls which couldn't be done with block supports

This is not possible now sure but after #29247 it will be possible to use the "block" segment for that. I'm fine if it's added to the end personally until that other PR lands.

File: packages/block-library/src/image/edit.js (hunk 5 above)
Comment (ajlende) [reply packages/block-library/src/image/edit.js:279]: The IDs here don't really need to be completely unique—they just need to be unique to the color combination.

I like the readability of the generated styles this way rather than a pseudo-random unique ID. When I demoed the early prototype internally at Automattic, I was using the block ID as a unique ID for the stylesheet, but I received feedback that it would be nice if the duplicate duotones could be combined into a single SVG and stylesheet for the page.

If we were doing something generic for outside usage, I think it might be better if we could just have some sort of hashing function rather than generating UUIDs so that we might still be able to do that deduplication eventually. Although, I'm still not really sure how that deduplication would work.

Comment (youknowriad) [reply packages/block-library/src/image/edit.js:279]: I personally think unique ids are better because conceptually this is an inline style specific to the current block.
The SVG is not though and yes, it would be good if we can dedup the SVGs (they are duplicated even now) but I don't think it makes sense to base the classname added to the block on the value because we'll be adding a classname per inline style while it's better to have a unique classname for the block and just apply on  the inline styles to the same classname.

File: packages/components/src/index.js (hunk 7 above)
Comment (ajlende) [reply packages/components/src/index.js:32]: The only reason I have it exported is because it seemed like it could be useful for others, and it was more convenient to export that than create a new `DuotonePalette` like is done for `ColorPalette`. So I'm guessing you'd prefer exporting a new `DuotonePalette` instead?

File: lib/duotone-filter.php (hunk 4 above)
Comment (ajlende) [reply lib/duotone-filter.php:16]: Yes, if we go with what you said in https://github.com/WordPress/gutenberg/pull/26751#discussion_r583430826, we can make that happen. 👍

File: packages/block-library/src/image/edit.js (hunk 5 above)
Comment (ajlende) [reply packages/block-library/src/image/edit.js:279]: > conceptually this is an inline style specific to the current block

This is very styling-centric, so conceptually for me, this is the same as using something like `hue-rotate` in CSS. If I was writing this HTML and CSS from scratch, I'd probably have a classname to apply the style, especially if I only have a couple different filters that I would want to apply to multiple images.

It also reminds me that it's possible to [inline the style within the CSS](https://css-tricks.com/svg-within-css/). If it were implemented that way, would you think differently about it?

File: packages/components/src/index.js (hunk 7 above)
Comment (youknowriad) [reply packages/components/src/index.js:32]: Yeah, I prefer exporting high level components more than lower level ones as it matches what we did before.

File: packages/block-library/src/image/edit.js (hunk 5 above)
Comment (youknowriad) [reply packages/block-library/src/image/edit.js:279]: > It also reminds me that it's possible to inline the style within the CSS. If it were implemented that way, would you think differently about it?

Potentially yes, but only for presets, for custom duotones, It seems that it's still better to me to have a unique class for the block. Basically, I see it as the same thing as an "inline style" applied to the element `style=""` since it's a custom value, the only difference is that we can't really use that because we need to target "sub items".

Comment (youknowriad) [reply packages/block-library/src/image/edit.js:279]: Thinking more here, I can see the point though about considering these as actually "presets" but they are just presets generated dynamically in a single  page/post. The problem with that is that right now, they are being generated in the block itself so they are block specific where they shouldn't be if we follow that principle.

So it's one of two things for me:

 - Embrace the fact that these are inline styles and use a unique id.
 - Consider these as "dynamic presets" and in that case don't inline the style but instead have a way to "register dynamic custom presets" which means we can dedupe as well.

I still lean towards 1 because it's how we approached things so far for block customizations, and I don't want us to introduce an API/behavior without thinking holistically about it and its impact.

File: packages/components/src/index.js (hunk 7 above)
Comment (ajlende) [reply packages/components/src/index.js:32]: I started refactoring it, but I'm having second guesses if that's the right way to go about this. 

Creating a new `DuotonePalette` component in `@wordpress/components` would be splitting the duotone logic between `@wordpress/components` and `@wordpress/block-editor`. [`getGradientFromCSSColors`](https://github.com/WordPress/gutenberg/pull/26751/files#diff-6621fb4a69f1cfc8ce0f2ea9bf7a88fca448c8ccc4ddc5ae1fb43c56738b480bR52-R68) and [`getValuesFromColors`](https://github.com/WordPress/gutenberg/pull/26751/files#diff-6621fb4a69f1cfc8ce0f2ea9bf7a88fca448c8ccc4ddc5ae1fb43c56738b480bR88-R107) would need to be used in both packages. Otherwise, if I go ahead and abstract out those parts to keep the duotone functions in `@wordpress/block-editor`, I'm basically just left with a light wrapper around `CircularOptionPicker`.

File: lib/duotone-filter.php (hunk 4 above)
Comment (ajlende) [reply lib/duotone-filter.php:16]: If this is getting moved exclusively to block supports, you can start reviewing the blocks supports parts unique to https://github.com/WordPress/gutenberg/pull/26752/files#diff-fd626ca219f3d805667d480eabb2d890a5a8d2e015b37f03cf8980edf1ae3d58 and https://github.com/WordPress/wordpress-develop/pull/984

**Comment by ellatrix** [comment]
What's the reasoning for doing this with SVG? Is it not possible to do with CSS filters?

**Comment by jasmussen** [comment]
> What's the reasoning for doing this with SVG? Is it not possible to do with CSS filters?

It's not possible. It's [possible to get close](https://codepen.io/joen/pen/BaoNgJw), but due to the way it's done, colors will always be inaccurate, darker and more muted. Whereas with SVG filters, the actual colors that make up the grayscale of an image can be change directly, allowing for [perfectly accurate color choices](https://codepen.io/joen/pen/OJPbJrX). You can even make total inversions.

File: packages/block-library/src/image/edit.js (hunk 5 above)
Comment (ajlende) [reply packages/block-library/src/image/edit.js:279]: > I see it as the same thing as an "inline style" applied to the element

I think I follow your reasoning now. It's more like an inline style because if you change it for one block, it's not going to affect the rest of the blocks that use the same color combination. A thing like "dynamic presets" would probably have to allow for updating everywhere when you update the preset, for example. Centralizing those things as "dynamic presets" would probably make deduplication easier too.

> Do you think we can build a generic thing for that and maybe extract it to its own PR first
 
I don't know if it really needs it. Is it good enough to just use the block id or `uuid()` like you have in #29335 instead? That's already pretty simple.

Comment (youknowriad) [reply packages/block-library/src/image/edit.js:279]: yes, I think we should start with just a uuid at first, (left some related comments on the block supports PR)

**Comment by ajlende** [comment]
Image block support moved to #26752. See https://github.com/WordPress/gutenberg/pull/26751#discussion_r583427663
